### PR TITLE
python311Packages.rapidocr-onnxruntime: init at 1.3.22

### LINKED ISSUE #317065: Package request: Python3Packages.rapidocr_onnxruntime
**Project description**

Awesome OCR multiple programing languages toolkits based on ONNXRuntime, OpenVION and PaddlePaddle.

**Metadata**

* homepage URL: https://github.com/RapidAI/RapidOCR
* source URL: https://github.com/RapidAI/RapidOCR/tree/main/python/rapidocr_onnxruntime
* license: asl20
* platforms: linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by pluiedev**
## Description of changes

Fixes #317065

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by drupol — COMMENTED** [review]
I think it should be moved under `pkgs/development/python-modules`. This is not an application, this is a module.

**Review by natsukium — DISMISSED** [review]
Please read the document first, especially the contributing guidelines section.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines-contributing-guidelines
And if necessary.
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#building-packages-and-applications-building-packages-and-applications

**Comment by pluiedev** [comment]
> Please read the document first, especially the contributing guidelines section. [`master`/doc/languages-frameworks/python.section.md#contributing-guidelines-contributing-guidelines](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md?rgh-link-date=2024-06-12T01%3A04%3A07Z#contributing-guidelines-contributing-guidelines) And if necessary. [`master`/doc/languages-frameworks/python.section.md#building-packages-and-applications-building-packages-and-applications](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md?rgh-link-date=2024-06-12T01%3A04%3A07Z#building-packages-and-applications-building-packages-and-applications)

I'm honestly not sure what I'm missing here — if you have any specific concerns, please just spell them out instead of making people guess

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:23]: Inherit those dependencies from top level parameters, not from `python3.pkgs`

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:23]: I guess this is more a personal preference thing, but sure

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:16]: I also think it needs a `pyImportCheck = [ "rapidocr_onnxruntime" ];`

**Comment by natsukium** [comment]
> I'm honestly not sure what I'm missing here — if you have any specific concerns, please just spell them out instead of making people guess. Telling people to RTFM is not particularly productive IMO

Are you going to contribute without even reading the contribution guide?
I'm not asking you to read all the documents. Just a checklist.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:48]:
```suggestion
```
Please use `pythonRelaxDepsHook` instead.
> The version restrictions of dependencies can be relaxed by [pythonRelaxDepsHook](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonrelaxdepshook).

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-contributing

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:52]:
```suggestion
  pythonImportsCheck = [ "rapidocr_onnxruntime" ];
```
> pythonImportsCheck is set. This is still a good smoke test even if pytestCheckHook is set.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-contributing

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:53]:
```suggestion
  nativeCheckInputs = [ pytestCheckHook ];
```
> Make sure the tests are enabled using for example [pytestCheckHook](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pytestcheckhook) and, in the case of libraries, are passing for all interpreters. If certain tests fail they can be disabled individually. Try to avoid disabling the tests altogether. In any case, when you disable tests, leave a comment explaining why.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-contributing

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:47]: Everything related to patches should be near the `src`.
It doesn't matter where it is as a processor, but the cognitive load should be low since humans do the maintenance.

**Comment by natsukium** [comment]
The commit message and PR title should be `python311Packages.rapidocr-onnxruntime`.
> Commit names of Python libraries should reflect that they are Python libraries, so write for example python311Packages.numpy: 1.11 -> 1.12. It is highly recommended to specify the current default version to enable automatic build by ofborg. Note that pythonPackages is an alias for python27Packages.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-contributing

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:52]: Oh my bad. Sorry for the misleading information.

**Comment by pluiedev** [comment]
> > I'm honestly not sure what I'm missing here — if you have any specific concerns, please just spell them out instead of making people guess. Telling people to RTFM is not particularly productive IMO
> 
> Are you going to contribute without even reading the contribution guide? I'm not asking you to read all the documents. Just a checklist.

Look, I have packaged Python packages before, and while I'm not quite up to speed on what's the expected standard of today, that's not a reason to just throw a big document at me and say "fix it." Please, just state problems directly.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:48]: This assumes the package requirements to be at `requirements.txt`, right? That doesn't work since the upstream file is at `requirements-ort.txt`.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:35]: I think the patch will require a comment explaining why it is needed.

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:35]: That's true — it's just that the script by default pulls the current version from PyPI and bumps it, which is obviously not allowed in the Nix sandbox

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:40]: Might as well rename the patch to something more descriptive I think — the current name is very vague

**Comment by pluiedev** [comment]
Ugh, looks like the tests use some models that are not bundled with the Git repo — how are you even supposed to get them, then?

**Comment by drupol** [comment]
> ModuleNotFoundError: No module named 'rapidocr_onnxruntime'

@natsukium Do you have an idea on how we could fix this?

**Comment by pluiedev** [comment]
> > ModuleNotFoundError: No module named 'rapidocr_onnxruntime'
> 
> @natsukium Do you have an idea on how we could fix this?

I suspect that the wheel isn't actually built correctly unless the models are bundled with it — I tried adding some of the models and the build process actually added those to the wheel, so that might be the reason why

**Comment by pluiedev** [comment]
Well, adding the models didn't fix the import check but at least now all tests work :tada:

**Comment by drupol** [comment]
Stupid question, what's `wheel` in this context?

**Comment by pluiedev** [comment]
> Stupid question, what's `wheel` in this context?

AFAICT a wheel is just the built binary Python package, in this case `rapidocr_onnxruntime-1.3.22-py3-none-any.whl`

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:31]: I'm wondering if this is a good idea...

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:73]: I guess this will have to be fixed before merging. What's the error log?

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:73]: Same as before: ModuleNotFoundError: No module named 'rapidocr_onnxruntime'.
Though, I think I found a reason for that — somehow the `__init__,py` and `main.py` are just completely missing from the built wheel. Given this project uses a very unconventional layout, I wonder if we're missing something again...

Comment (natsukium) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:73]: This is a blocker.
```shell
$ nix-shell -p 'python311.withPackages (ps: with ps; [rapidocr-onnxruntime])' -I nixpkgs=. --run 'python -c "import rapidocr_onnxruntime"'

Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'rapidocr_onnxruntime'
```

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:73]: Well, of course it's a blocker, as it doesn't even work :sweat_smile:

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:31]: If it is used for testing only, it is not necessary.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:85]: Regarding `__init__.py`, we need to add the following command to `postPatch`.
```
echo "from .rapidocr_onnxruntime.main import RapidOCR, VisRes" > rapidocr_onnxruntime/__init__.py
```
https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:85]: That doesn't quite work with the current setup — tests fail:
```
       > ______________________ ERROR collecting tests/test_ort.py ______________________
       > ImportError while importing test module '/build/source/python/tests/test_ort.py'.
       > Hint: make sure your test modules/packages have valid Python names.
       > Traceback:
       > /nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/importlib/__init__.py:126: in import_module
       >     return _bootstrap._gcd_import(name[level:], package, level)
       > tests/test_ort.py:16: in <module>
       >     from rapidocr_onnxruntime import LoadImageError, RapidOCR
       > rapidocr_onnxruntime/__init__.py:1: in <module>
       >     from .rapidocr_onnxruntime.main import RapidOCR, VisRes
       > E   ModuleNotFoundError: No module named 'rapidocr_onnxruntime.rapidocr_onnxruntime'
```

Or with `from .rapidocr_onnxruntime.main import RapidOCR, VisRes`:
```
       > ______________________ ERROR collecting tests/test_ort.py ______________________
       > ImportError while importing test module '/build/source/python/tests/test_ort.py'.
       > Hint: make sure your test modules/packages have valid Python names.
       > Traceback:
       > /nix/store/7hnr99nxrd2aw6lghybqdmkckq60j6l9-python3-3.11.9/lib/python3.11/importlib/__init__.py:126: in import_module
       >     return _bootstrap._gcd_import(name[level:], package, level)
       > tests/test_ort.py:16: in <module>
       >     from rapidocr_onnxruntime import LoadImageError, RapidOCR
       > E   ImportError: cannot import name 'LoadImageError' from 'rapidocr_onnxruntime' (/build/source/python/rapidocr_onnxruntime/__init__.py)
```

Comment (natsukium) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:85]: I suspect the `setuptools` namespace is causing the problem.
https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/python/setup_onnxruntime.py#L59-L60

https://setuptools.pypa.io/en/latest/userguide/package_discovery.html

Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:85]: > Regarding `__init__.py`, we need to add the following command to `postPatch`.
> 
> ```
> echo "from .rapidocr_onnxruntime.main import RapidOCR, VisRes" > rapidocr_onnxruntime/__init__.py
> ```
> 
> https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87

Why god why are they doing that ?!

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:70]:
```suggestion
  preBuild = ''
    mkdir rapidocr_onnxruntime_t
    mv rapidocr_onnxruntime rapidocr_onnxruntime_t
    mv rapidocr_onnxruntime_t rapidocr_onnxruntime
  '';
```
I don't know what they want to do.
`pythonImportsCheck` now passes, but it still seems to be missing some files during the test.

https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L83-L85

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:66]: `pytestCheckHook` should go to `nativeCheckInputs`.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:70]: before
```shell
$ cat /tmp/nix-build-python3.11-rapidocr-onnxruntime-1.3.22.drv-2/source/python/rapidocr_onnxruntime/__init__.py 
from .rapidocr_onnxruntime.main import RapidOCR, VisRes
```

after
```shell
$ cat /tmp/nix-build-python3.11-rapidocr-onnxruntime-1.3.22.drv-3/source/python/rapidocr_onnxruntime/rapidocr_onnxruntime/__init__.py 
# -*- encoding: utf-8 -*-
# @Author: SWHL
# @Contact: liekkaskono@163.com
from .main import RapidOCR
from .utils import LoadImageError, VisRes
```

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:86]:
```suggestion
  preCheck = ''
    mv rapidocr_onnxruntime rapidocr_onnxruntime_t
    mv rapidocr_onnxruntime_t/* .
  '';
```
It's a dirty hack.
We can create a directory structure `rapidocr_onnxruntime/rapidocr_onnxruntime` before the build and restore it after the build.

**Comment by natsukium** [comment]
In short, we need three things.
1. create `__init__.py` in `postPatch` phase [ref](https://github.com/NixOS/nixpkgs/pull/319053#discussion_r1637848253)
2. move `rapidocr_onnxruntime` to `rapidocr_onnxruntime/rapidocr_onnxruntime` in `preBuild` phase [ref](https://github.com/NixOS/nixpkgs/pull/319053#discussion_r1637927181)
3. revert the operation 2 in `preCheck` phase (maybe `postBuild` phase is ok) [ref](https://github.com/NixOS/nixpkgs/pull/319053#discussion_r1637955272)

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:85]: > > Regarding `__init__.py`, we need to add the following command to `postPatch`.
> > ```
> > echo "from .rapidocr_onnxruntime.main import RapidOCR, VisRes" > rapidocr_onnxruntime/__init__.py
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [RapidAI/RapidOCR@`02829ef`/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87](https://github.com/RapidAI/RapidOCR/blob/02829ef986bc2a5c4f33e9c45c9267bcf2d07a1d/.github/workflows/gen_whl_to_pypi_rapidocr_ort.yml#L86-L87)
> 
> Why god why are they doing that ?!

Beats me... This package is just built in an unorthodox way in every sense of the word.

**Comment by natsukium** [comment]
In general, machine learning engineers are only interested in their own environment.
I know this well because I am also a machine learning engineer.

**Comment by pluiedev** [comment]
Right, makes sense... should we mark this ready for review then?

**Comment by natsukium** [comment]
Yes, please.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:70]: Could you move this between `pythonImportsCheck` and `disabledTestPaths`?

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:86]: `opencv-python` is not a vendor package, it is a standard package called opencv-python in pypi.
We need `pythonRemoveDeps` because we have adopted the `opencv4` as an attribute name.https://pypi.org/project/opencv-python/

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:112]: Could you move `postPatch` after `patches`?

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:126]: These are related to the build phase, so it is preferred to be near `nativeBuildInputs` or `build-system`.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:130]:
```suggestion
    homepage = "https://github.com/RapidAI/RapidOCR";
    changelog = "https://github.com/RapidAI/RapidOCR/releases/tag/v${version}";
```
Adding meta.changelog is useful for future updates.

**Comment by drupol** [comment]
@pluiedev Hope you don't mind me pushing a commit, it is mainly for helping you. I didn't rebased it myself so it's easier to pickup.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix:132]: Fixed.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix]: Fixed.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix]: Fixed.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix]: Fixed.

File: pkgs/development/python-modules/rapidocr-onnxruntime/default.nix
Comment (drupol) [reply pkgs/development/python-modules/rapidocr-onnxruntime/default.nix]: Fixed.

**Comment by natsukium** [comment]
on aarch64-linux🤔
something is wrong with the import check.
https://github.com/NixOS/nixpkgs/pull/319053/checks?check_run_id=26206549989

**Comment by drupol** [comment]
```
       > Error in cpuinfo: failed to parse the list of possible processors in /sys/devices/system/cpu/possible
       > Error in cpuinfo: failed to parse the list of present processors in /sys/devices/system/cpu/present
       > Error in cpuinfo: failed to parse both lists of possible and present processors
       > terminate called after throwing an instance of 'onnxruntime::OnnxRuntimeException'
```

**Comment by pluiedev** [comment]
Can repro by running `nix build .#legacyPackages.aarch64-linux.python3Packages.rapidocr-onnxruntime`. This looks like perhaps an OnnxRuntime issue...?

**Comment by pluiedev** [comment]
With further digging I realized what's going wrong here — on aarch64, ONNX Runtime uses pytorch's cpuinfo to look for CPU core configuration, by looking at info provided by `/sys/devices/system/cpu/{present,possible}`. However, that information is not exposed in the Nix builder, possibly to disallow builders from relying on the build machine's specific hardware configuration that may result in software that doesn't work for other systems. See pytorch/cpuinfo#143, microsoft/onnxruntime#10038.

In my mind there's two ways going forward — either mark aarch64 builds as broken, or patch ONNX Runtime to build without using cpuinfo by disabling the CMake flag [`onnxruntime_ENABLE_CPUINFO`](https://github.com/microsoft/onnxruntime/blob/983fd8393aa30b0c788275ea6e513a614274e1ba/cmake/CMakeLists.txt#L229).

**Comment by drupol** [comment]
IMHO, we should disable [onnxruntime_ENABLE_CPUINFO](https://github.com/microsoft/onnxruntime/blob/983fd8393aa30b0c788275ea6e513a614274e1ba/cmake/CMakeLists.txt#L229) on `aarch64`. Anyone has another opinion?

**Comment by pluiedev** [comment]
Agreed — working slowly is better than not working at all! :P

**Comment by pluiedev** [comment]
I did some more work to package the OpenVINO and PaddlePaddle versions of this package too — wasn't that hard

File: pkgs/development/python-modules/rapidocr/default.nix
Comment (drupol) [inline pkgs/development/python-modules/rapidocr/default.nix:1]: I don't think this is allowed in `nixpkgs`.

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr/default.nix:1]: Perhaps, or just don't supply a default to make it fail if it's not called via callPackage with the right arg

Comment (drupol) [reply pkgs/development/python-modules/rapidocr/default.nix:1]: @natsukium What do you recommend here?

Comment (natsukium) [reply pkgs/development/python-modules/rapidocr/default.nix:1]: We should have one file per package.
The python library is often broken, so we set attributes like `disabled` and `meta.broken`, and I don't want to put another condition, backend. `paddlepaddle` is an example.

Monorepo packages, for example `opentelemetry`, could share the same version, src, etc.
It may be more duplicated than the current programmatic one, and it may be uglier, depending on your preference, but for future maintenance, this is the way to go.
https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/opentelemetry-api/default.nix
https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/opentelemetry-sdk/default.nix

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr/default.nix:1]: Honestly the approach showed above isn't very DRY and requires a lot of duplicated code, or sourcing code from one place only to change it later. Since 95% of the derivation is exactly the same across all backends, I find that approach adds a needless burden on maintainers who have to make changes to all three runtimes once upstream changes something.

Comment (drupol) [reply pkgs/development/python-modules/rapidocr/default.nix:1]: How about merging `rapidocr-onnxruntime` for the moment which seems to be in good state, and decide what to do after that? I'm badly looking for that package in `open-webui` actually.

Comment (pluiedev) [reply pkgs/development/python-modules/rapidocr/default.nix:1]: Sure!

**Comment by drupol** [comment]
Should we patch `onnxruntime` for `aarch64-linux` ?

**Comment by drupol** [comment]
@natsukium @pluiedev Do you think it's ready for merge or... ?

**Comment by drupol** [comment]
@ofborg build python3Packages.rapidocr-onnxruntime

**Comment by drupol** [comment]
OK Planning to merge this in 1 or 2 hours.

**Pull request merged**
